---
title: "Yeast expression"
author: "Lavrentii Danilov"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache = TRUE)
```

```{r libraries, include=FALSE}
library(readxl)
library(limma)
library(ape)
library(dendextend)
library(RColorBrewer)
library(pvclust)
library(gplots)
library(NMF)
library(vegan)
library(Biobase)
library(DT)
library(ggplot2)
library(impute)
library(ggrepel)
```

# Data upload.

```{r include=FALSE}
yeast <- read.csv('report.unique_genes_matrix.tsv', sep='\t')
rownames(yeast) <- yeast$Genes
yeast <- yeast[,-1]
yeast <- yeast[, !names(yeast) %in% c('X2_wt_2', 'X2_wt_3')]
```

## EDA


```{r}
genes_with_NA <- names(which(rowSums(is.na(yeast)) > ncol(yeast)/2))
yeast <- yeast[!rownames(yeast) %in% genes_with_NA,]
```

Let's make an imputation of missing values in the remaining genes by KNN method.

```{r}
yeast <- as.matrix(yeast)
yeast_trans <- t(yeast)
knn_yeast <- impute.knn(yeast_trans, k = 5)
knn_yeast <- knn_yeast$data
knn_yeast <- as.data.frame(knn_yeast)
#knn_yeast$LEU2 <- knn_yeast$LEU2 / 2
#knn_yeast$URA3 <- knn_yeast$URA3 / 2
knn_yeast <- t(as.matrix(knn_yeast))

#yeast_experiment <- as.factor(c("wt_1", "wt_1", "wt_1", "wt_1", "105_2", "105_2", "105_2", "105_2", "wt_2", "wt_2", "105_3", "105_3", "105_3", "105_3", "wt_3", "wt_3", "wt_3", "wt_3"))

yeast_experiment <- as.factor(c("1", "1", "1", "1", "2", "2", "2", "2"))
```

```{r}
pal <- brewer.pal(n = 9, name = "Set1")
cols <- pal[yeast_experiment]
boxplot(knn_yeast, outline = FALSE, main = "Raw data", col = cols)
legend("topright", levels(yeast_experiment), fill = pal, bty = "n", xpd = T)
```

Do data normalization -- logarithm + quantile normalization

```{r}
yeast_log <- log2(knn_yeast)
yeast_norm <- normalizeQuantiles(as.matrix(yeast_log))
boxplot(yeast_norm, outline = FALSE, main = "Normalized data", col = cols)
legend("topright", levels(yeast_experiment), fill = pal, bty = "n", xpd = T)
```

###PCA
```{r}
yeas_pca <- t(yeast_norm)
terapod_pca <- rda(yeas_pca, scale = TRUE)

rownames(yeas_pca) <- c("105_3", "105_3", "105_3", "105_3", "wt_3", "wt_3", "wt_3", "wt_3")

df_scores <- data.frame(yeas_pca,
                        scores(terapod_pca, display = "sites", choices = c(1, 2, 3), scaling = "sites"))

p_scores <- ggplot(df_scores, aes(x = PC1, y = PC2)) + 
  geom_point(aes(color = rownames(yeas_pca)), alpha = 0.5) +
  coord_equal(xlim = c(-3, 3), ylim = c(-3, 3)) + ggtitle(label = "Ordination in the axes of principal components") + theme_bw()
p_scores
```

###Differential expression 

```{r}
# Данные экспрессии
expr_data <- as.matrix(yeast_norm)

# Данные о пробах
pheno_data <- data.frame(yeast_experiment)
rownames(pheno_data) <- colnames(yeast_norm)
pheno_metadata <- data.frame(
  labelDescription = c("Experimental condition"),
  row.names=c("Condition"))
pheno_data <- new("AnnotatedDataFrame",
                 data = pheno_data,
                 varMetadata = pheno_metadata)

# Данные о признаках (белках)
feature_data <- data.frame(Prot = rownames(expr_data))
rownames(feature_data) <- rownames(expr_data)
feature_metadata <- data.frame(
  labelDescription = c("Protain name"),
  row.names = c("Protain"))
f_data <- new("AnnotatedDataFrame",
              data = feature_data,
              varMetadata = feature_metadata)

# Собираем вместе
exp_set <-
  ExpressionSet(assayData = expr_data,
                phenoData = pheno_data,
                featureData = f_data)
```


```{r}
X <- model.matrix(~ yeast_experiment, pData(exp_set))
fit <- lmFit(exp_set, design = X, method = "robust", maxit = 1000)
efit <- eBayes(fit)
```


```{r}
MA_limma <- function(efit, coef, n = 10, signif = TRUE, fdr = 0.05, lfc = 0, text = TRUE, cex.text = 0.8, col.text = "grey20", main = "MA-plot", xlab = "Average log-expression", ylab = "Expression log-ratio", pch = 19, pch.signif = 21, col = "darkgreen", alpha = 0.3, cex = 0.3, ...){
  R <- efit$coefficients[, coef]
  I <- efit$Amean
  col_btransp <- adjustcolor(col, alpha.f = alpha)
  plot(I, R, cex = cex, main = main, pch = pch, xlab = xlab, ylab = ylab, col = col_btransp, ...)
  abline(h = 0)
  if(signif){
    sign <- p.adjust(efit$p.value[, coef], method = "BH") <= fdr
    large <- abs(efit$coefficients[, coef]) >= lfc
    points(I[sign & large], R[sign & large], cex = cex*2, col = "orange2", pch = pch.signif)
  }
  if(text){
    ord <- order(efit$lods[, coef], decreasing = TRUE)
    top_n <- ord[1:n]
    text(I[top_n], R[top_n], labels = efit$genes[top_n, ], pos = 4, cex = cex.text, col = col.text)
  }
}


MA_limma(efit, coef = 2, n = 30)
```


```{r}
my_list <- topTable(efit, coef = 2, n = 100)
dif_exp_set <- exp_set[fData(exp_set)$Prot %in% my_list$Prot, ]
```


```{r fig.height=10, fig.width=15}
dat <- as.matrix(exprs(dif_exp_set))
# после дополнительной стандартизации по белкам
pal_blue_red <- colorpanel(75, low = "steelblue", mid = "black", high = "red")
heatmap.2(dat, col = pal_blue_red, scale = "row", key = TRUE, symkey = FALSE, density.info = "none", trace = "none", cexRow = 0.9, cexCol = 1, margins = c(4, 3), keysize = 0.8, key.par = list(mar = c(3, 0.1, 3, 0.1)))
```

```{r}
topTable(efit, coef = 2)
numGenes <- nrow(exprs(exp_set))
full_list <- topTable(efit, number = numGenes)
#full_list <- full_list[full_list$adj.P.Val <= 0.05,]
write.csv(full_list, 'DE_Proteomics_full.csv')
```

# Volcano plot 

```{r }
volcano_list <- full_list

volcano1 <- ggplot(data = volcano_list, aes(x = yeast_experiment1, y = -log10(adj.P.Val))) + geom_point()

volcano2 <- volcano1 + geom_vline(xintercept=c(-0.6, 0.6), col="red") +
  geom_hline(yintercept=-log10(0.05), col="red")

volcano_list$diffexpressed <- "NO"       
```


```{r }
volcano_list$diffexpressed[volcano_list$yeast_experiment1 > 2 & volcano_list$adj.P.Val < 0.05] <- "UP"
```

 if log2Foldchange < -2 and pvalue < 0.05, set as "DOWN"

```{r }
volcano_list$diffexpressed[volcano_list$yeast_experiment1 < -2 & volcano_list$adj.P.Val< 0.05] <- "DOWN"

volcanodif1 <- ggplot(data = volcano_list, aes(x = yeast_experiment1, y = -log10(adj.P.Val), col = diffexpressed)) + geom_point() + theme_minimal()
```

 Add lines as before...

```{r }
volcanodif2 <- volcanodif1 + geom_vline(xintercept=c(-2, 2), col="red") +
  geom_hline(yintercept=-log10(0.05), col="red")

volcano_list$Prot <- as.character(volcano_list$Prot)
volcano_list$delabel <- NA
volcano_list$delabel[volcano_list$diffexpressed != "NO"] <- volcano_list$Prot[volcano_list$diffexpressedxpressed != "NO"]
#volcano_list[volcano_list$adj.P.Val<=0.6,]$delabel <- volcano_list[volcano_list$adj.P.Val<=0.6,]$Prot

plot_proteins <- ggplot(data=volcano_list, aes(x=yeast_experiment1, y=-log10(adj.P.Val), col=diffexpressed, label=delabel)) + 
  geom_point() + 
  theme_minimal() +
  geom_text_repel(size=3, colour = 'black', max.overlaps = 30)

plot_final <- plot_proteins + geom_vline(xintercept=c(-2, 2), col="red") + geom_hline(yintercept=-log10(0.05), col="red")
#plot_final
```

### Tree and ordination by dif-expressed proteins

```{r fig.height=7, fig.width=15, message=FALSE, warning=FALSE}
diff_prot <- rownames(full_list)
diff_expressed <- as.data.frame(yeast_norm)[diff_prot,]
t_diff_expressed <- t(diff_expressed)
#rownames(t_diff_expressed) <-  as.factor(gsub("_[^_]*", replacement = "", rownames(t_diff_expressed)))
#rownames(t_diff_expressed) <- make.unique(as.character(pheno$yeast_experiment))

d <- dist(x = t_diff_expressed, method = "euclidean")

mouse_hc_avg <- hclust(d, method = "average")
mouse_ph_avg <- as.phylo(mouse_hc_avg)
mouse_den_avg <- as.dendrogram(mouse_hc_avg)

get_colours <- function(dend, n_chars, palette = "Dark2"){ #nchars = первые нескольок симовлов которые будут использоваться для разделения фактора на группы
labs <- get_leaves_attr(dend, "label")
group <- substr(labs, start = 0, stop = n_chars)
group <- factor(group)
cols <- brewer.pal(length(levels(group)), name = palette)[group]
return(cols)
}

cols <- get_colours(dend = mouse_den_avg, n_chars = 6)
den_avg_c <- color_labels(dend = mouse_den_avg, col = cols)
plot(den_avg_c, horiz = TRUE)
```
